### PR TITLE
[Parse] Reapply trailing closure in condition fix and downgrade to warning

### PR DESCRIPTION
**Pull request by owenv**
This reapplies #25165 and brings it up to date. It also downgrades the error to a warning to preserve source compatibility.

### CODE REVIEW

**Comment by owenv** [comment]
@rintaro As Jordan suggested in the previous PR, this conservatively downgrades the error to a warning even in cases where we were previously diagnosing the condition correctly. Ideally, I'd like to change the warning text to `trailing closure requires parentheses for disambiguation in this context; this will become an error in a future language version`, and then upgrade this to an error next time a new source-breaking language mode is introduced. I wanted to get your opinion first though since I think that'd need core team approval.

**Comment by rintaro** [comment]
I don't even think we should change it to an error in the future. This is a kind of `do XXX to silence this warning` diagnostics.
How about `warning: trailing closure in this context is confusable with the body of the statement; do not use trailing closure to silence this warning` or such?

**Comment by owenv** [comment]
@rintaro Sounds good to me! I changed it to `trailing closure in this context is confusable with the body of the statement; pass as a parenthesized argument to silence this warning`. I think `pass as a parenthesized argument` describes the fix-it a little better, although it is a little long.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by rintaro — APPROVED** [review]
Looks great. Thank you!

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: include/swift/AST/DiagnosticsSema.def
Comment (rintaro) [inline include/swift/AST/DiagnosticsSema.def]:
```suggestion
        "trailing closure in this context is confusable with the body of the "
```
Could you update?

Comment (owenv) [reply include/swift/AST/DiagnosticsSema.def]: Sorry about that, must have snuck in when I wrapped the line to 80 chars! It should be fixed now

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
Thank you Owen!

**Pull request merged**
